File: w3/w3cpp/Pen_Paper/Paper.cpp
#include <iostream>
#include "Paper.h"

Paper::Paper(int maxSymbols) {
    this->maxSymbols = maxSymbols;
    this->symbols = 0;
}

Paper::~Paper() {
    
}

int Paper::getMaxSymbols() const {
    return this->maxSymbols;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const std::string& message) {
    int spaceNeeded = message.size();
    int spaceAmount = this->maxSymbols - this->symbols;
    
    if ( this->symbols == this->maxSymbols ) {
        throw OutOfSpace();
    }
    
    if ( spaceAmount < spaceNeeded ) {
        spaceNeeded = spaceAmount;
    }
    
    this->symbols += spaceNeeded;
    
    content += message.substr(0, spaceAmount);
}

void Paper::show() const {
    std::cout << content << std::endl;
}

File: w3/w3java/Pen_Paper/Paper.java
public class Paper {
    private int maxSymbols;
    private int symbols;
    private StringBuffer content;
    
    public Paper(int maxSymbols) {
        this.maxSymbols = maxSymbols;
        this.symbols = 0;
        content = new StringBuffer();
    }
    
    public Paper() {
        this.maxSymbols = 4096;
        this.symbols = 0;
        content = new StringBuffer();
    }
    
    public int getMaxSymbols() {
        return this.maxSymbols;
    }
    
    public int getSymbols() {
        return this.symbols;
    }
    
    public void addContent(String message) throws OutOfSpaceException {
        int total = content.length() + message.length();
        
        if ( content.length() == this.maxSymbols ) {
            throw new OutOfSpaceException();
        }
        
        if ( total > this.maxSymbols ) {
            content.append(message.substring(0, this.maxSymbols-content.length()));
            return;
        }
        
        content.append(message);
    }
    
    public void show() {
        System.out.println(content);
    }
}